File: day2/day2.hs
{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
data Instruction = Instruction {instructionAction :: String, instructionValue::  Int}

instructionParserLine :: String ->  Instruction
instructionParserLine line = Instruction (head (words line)) (read (words line !! 1) :: Int)

instructionParser :: String ->  [Instruction]
instructionParser inp = map instructionParserLine $ lines inp


applyStepPartA :: [Instruction] -> Int -> Int -> Int
applyStepPartA instructions dir dep = prod
    where
        prod = if null instructions
               then dir * dep
               else 
                case instructionAction (head instructions) of 
                    "forward" -> applyStepPartA (drop 1 instructions) (dir + instructionValue (head instructions)) dep
                    "up" -> applyStepPartA (drop 1 instructions) dir (dep - instructionValue (head instructions))
                    "down" -> applyStepPartA (drop 1 instructions) dir (dep + instructionValue (head instructions))

applyStepPartB :: [Instruction] -> Int -> Int -> Int -> Int
applyStepPartB instructions dir dep aim = prod
    where
        prod = if null instructions
               then dir * dep
               else 
                case instructionAction (head instructions) of 
                    "forward" -> applyStepPartB (drop 1 instructions) (dir + instructionValue (head instructions)) (dep + (aim * instructionValue (head instructions))) aim
                    "up" -> applyStepPartB (drop 1 instructions) dir dep (aim - instructionValue (head instructions)) 
                    "down" -> applyStepPartB (drop 1 instructions) dir dep (aim + instructionValue (head instructions))


day2 = do 
    parta <- readFile "data.txt"
    let parsedList = instructionParser parta
    let out_a = applyStepPartA parsedList 0 0
    print out_a
    let out_b = applyStepPartB parsedList 0 0 0
    print out_b

File: day2/day2.py

def p1(cords):
	start_pos = [0,0]

	for cord in cords:
		if cord[0] == 'forward':
			start_pos[0] += cord[1]
		elif cord[0] == 'up':
			start_pos[1] -= cord[1]
		elif cord[0] == 'down':
			start_pos[1] += cord[1]
		else:
			raise Exception(cord[0])
	print(start_pos[0] * start_pos[1])

def p2(cords):
	start_pos = [0,0,0]

	for cord in cords:
		if cord[0] == 'forward':
			start_pos[0] += cord[1]
			start_pos[1] += start_pos[2] * cord[1]
		elif cord[0] == 'up':
			start_pos[2] -= cord[1]
		elif cord[0] == 'down':
			start_pos[2] += cord[1]
		else:
			raise Exception(cord[0])
	print(start_pos[0] * start_pos[1])

if __name__ == '__main__':
	cords = [(lines.split(' ')[0], int(lines.split(' ')[1])) for lines in open("data.txt").readlines()]
	print("Part 1:")
	p1(cords)
	print("Part 2:")
	p2(cords)